Strange/funny error message with `include_bytes!`

The "fix-it" hint when using the `include_bytes!` macro where a `&str` is expected is rather interesting:

```
error[E0308]: mismatched types
 --> src/main.rs:1:21
  |
1 | static TEST: &str = include_bytes!("test.txt");
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |                     |
  |                     expected str, found array of 3 elements
  |                     help: consider removing the leading `b`: `nclude_bytes!("test.txt")`
  |
  = note: expected type `&'static str`
             found type `&'static [u8; 3]`
```

I was able to reproduce this on stable 1.24.0 and nightly.  Here are the relevant `rustc --version --verbose` outputs:

Stable:
```
rustc 1.24.0 (4d90ac38c 2018-02-12)
binary: rustc
commit-hash: 4d90ac38c0b61bb69470b61ea2cccea0df48d9e5
commit-date: 2018-02-12
host: x86_64-unknown-linux-gnu
release: 1.24.0
LLVM version: 4.0
```

Nightly:
```
rustc 1.25.0-nightly (27a046e93 2018-02-18)
binary: rustc
commit-hash: 27a046e9338fb0455c33b13e8fe28da78212dedc
commit-date: 2018-02-18
host: x86_64-unknown-linux-gnu
release: 1.25.0-nightly
LLVM version: 6.0
```

I don't know anything about the compiler internals, but my best guess is that the fix-it suggestion is generated based on the expanded macro output (which starts with a `b`), but the suggested "fix" is generated from the source text without knowing that the two are different.  Not a serious problem, but it's certainly entertaining :)  I tried searching the issue tracker to see if anyone else had experienced this, but I couldn't find anything, so my apologies if this is a duplicate of something.

EDIT: Right after filing the issue, I decided to go and track it down in the source: it originates [here](https://github.com/rust-lang/rust/blob/9fd7da904b46ff7aa78c2e2cc1986c4975aeccc6/src/librustc_typeck/check/demand.rs#L221), and could probably be fixed easily by checking if the first character is actually a `b` before outputting the message.  Not so sure about the error suggested by the code that immediately follows, which will add a `b` to `include_str!` if a `&[u8]` is expected.

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -264,9 +264,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 (&ty::TyStr, &ty::TySlice(arr)) if arr == self.tcx.types.u8 => {
                     if let hir::ExprLit(_) = expr.node {
                         if let Ok(src) = cm.span_to_snippet(sp) {
-                            return Some((sp,
-                                         "consider removing the leading `b`",
-                                         src[1..].to_string()));
+                            if src.starts_with("b\"") {
+                                return Some((sp,
+                                             "consider removing the leading `b`",
+                                             src[1..].to_string()));
+                            }
                         }
                     }
                 },
@@ -274,9 +276,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 (&ty::TySlice(arr), &ty::TyStr) if arr == self.tcx.types.u8 => {
                     if let hir::ExprLit(_) = expr.node {
                         if let Ok(src) = cm.span_to_snippet(sp) {
-                            return Some((sp,
-                                         "consider adding a leading `b`",
-                                         format!("b{}", src)));
+                            if src.starts_with("\"") {
+                                return Some((sp,
+                                             "consider adding a leading `b`",
+                                             format!("b{}", src)));
+                            }
                         }
                     }
                 }

--- a/src/test/ui/suggestions/issue-48364.rs
+++ b/src/test/ui/suggestions/issue-48364.rs
@@ -1,0 +1,16 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn foo() -> bool {
+    b"".starts_with(stringify!(foo))
+    //~^ ERROR mismatched types
+}
+
+fn main() {}

--- a/src/test/ui/suggestions/issue-48364.stderr
+++ b/src/test/ui/suggestions/issue-48364.stderr
@@ -1,0 +1,12 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-48364.rs:12:21
+   |
+LL |     b"".starts_with(stringify!(foo))
+   |                     ^^^^^^^^^^^^^^^ expected slice, found str
+   |
+   = note: expected type `&[u8]`
+              found type `&'static str`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

